updated to TypeScript and added acr policy propagation (#144)
solid/authorization-panel: proposals/acp/policy-propagation.md

```TypeScript
////////////////// RESOURCES //////////////////

/**

>  * Note this this function creates a Resource and therefore there is no

Reviewer: ```suggestion
 * Note that this function creates a Resource and therefore there is no
```

 * Policy propagation required. Propagation only occurs when Policy changees
 * happen on a Container.
 *
 * agent     : The Agent trying to PUT the resource
 * resource  : The Resource the Agent is trying to PUT
 * container : The parent Container for the Resource
 * content   : The content for the Resource the Agent wants to PUT in the Container
 *
 * Return    : 403 or 401 if the Agent does not have permission.
 *             204 if the resource already exists and if overwritten successfully.
 *             201 if the resource did not already exists and is created.
 */
export function putResource(agent: Agent, resource: Resource, container: Container, content: Readable) {
  if (resource.exists()) {
    if (!isAccessAllowed(agent, resource, HTTPMethod.PUT)) return accessError(agent);

    resource.write(content);
    return StatusCodes.NO_CONTENT;
  }

  if (!isAccessAllowed(agent, container, HTTPMethod.PUT)) return accessError(agent);

  createResource(container, resource.getIri(), content);

  return StatusCodes.CREATED;
}

/**
 * When an Agent sends a POST request to a container,
 * a new child Resource will be created in the container.
 *

>  * Note this this function creates a Resource and therefore there is no

Reviewer: ```suggestion
 * Note that this function creates a Resource and therefore there is no
```

 * Policy propagation required. Propagation only occurs when Policy changees
 * happen on a Container.
 *
 * agent     : The Agent trying to POST to the Container
 * container : The Container the Agent sent the request to
 * slug      : The hint provided by the requestor for the child resource name
 * content   : The content for the Resource the Agent wants to POST to the Container
 *
 * Return    : 403 or 401 if the Agent does not have permission.
 *             201 if the resource is created.
 */
export function postResource(agent: Agent, container: Container, slug: string, content: Readable) {
  if (!isAccessAllowed(agent, container, HTTPMethod.POST)) return accessError(agent);

  const iri: IRI = container.newResourceName(slug);
  createResource(container, iri, content);

  return StatusCodes.CREATED;
}

/**
 * Note this this function updates a Resource and therefore there is no
 * Policy propagation required. Propagation only occurs when Policy changees
 * happen on a Container.
 *

>  * If the PATCH only contains content to be appended (e.g. triples to add)

Reviewer: ```suggestion
 * If the PATCH only contains content to be appended (e.g., triples to add)
```

 * then acp:Append access is requred.
 * However, if the PATCH contains content that requires changes

>  * (e.g. triples to be removed) then acp:Write access is required.

Reviewer: ```suggestion
 * (e.g., triples to be removed), then acp:Write access is required.
```

 *

>  * agent    : The Agent tryping to PATH the resource.

Reviewer: ```suggestion
 * agent    : The Agent trying to PATCH the resource.
```

 * resource : The Resource the Agent is trying to PATCH
 * content  : The content the Agent is using to PATCH the Resource. This pseudo-code does

>  *            not attemptto describe how the patch operation is carried out.

Reviewer: ```suggestion
 *            not attempt to describe how the patch operation is carried out.
```

 *
 * Return   : 403 or 401 if the Agent does not have permission.
 *            204 if the resource is patched successfully.
 *
 */
export function patchResource(agent: Agent, resource: Resource, content: Readable, appendOnly: boolean) {
  if (appendOnly) {
    if (!isAccessAllowed(agent, resource, HTTPMethod.PATCH)) return accessError(agent);
  } else {
    if (!isAccessAllowed(agent, resource, HTTPMethod.PUT)) return accessError(agent);
  }
  if (!resource.exists()) return StatusCodes.NOT_FOUND;

  resource.applyPatch(content);

  return StatusCodes.NO_CONTENT;
}

/**
 * Delete a resource and the associated ACR.
 *
 * NOTE : FOR DISCUSSION. I think it would be better if we had a specific DELETE
 *        access rather than overloading the WRITE access. This would be on the
 *        resource directly rather than on the parent container. By overloading the

>  *        Write access we are assuming there are no use cases where somebody should not

Reviewer: ```suggestion
 *        Write access, we are assuming there are no use cases where somebody should not
```

 *        be able to delete a resource even if they can remove the content. This
 *        would also map much cleaner to the HTTP methods.
 *        Deleting a resource does modify the containment triples in the container
 *        but I don't think that is a good reason to require acp:write access on
 *        the container because those triples are server managed and cannot be
 *        manipulated directly. Even if the decision was to still require acp:write

>  *        access on the container becuase it is changing, I still think it is a valid

Reviewer: ```suggestion
 *        access on the container because it is changing, I still think it is a valid
```

 *        case to have acp:delete to control deletion of the resource.
 */
export function deleteResource(agent: Agent, resource: Resource, container: Container) {
  if (!isAccessAllowed(agent, resource, HTTPMethod.DELETE)) return accessError(agent);
  if (!resource.exists()) return StatusCodes.NOT_FOUND;

  container.delete(resource);
  container.deleteACR(resource);
  container.deleteContainmentTriples(resource);

  return StatusCodes.NO_CONTENT;
}

////////////////// CONTAINERS //////////////////

/**
 * agent      : The Agent trying to PUT the container
 * container  : The Container the Agent is trying to PUT
 * pContainer : The parent Container for the container
 * content    : The content for the Container the Agent wants to PUT in the parent Container.

>  *              The content cannot includecontainment triples but can be any triples that are

Reviewer: ```suggestion
 *              The content cannot include containment triples but can be any triples that are
```

 *              valid in a normal Resource.
 *
 *              FOR DISCUSSION. If containment triples exist should they be ignored

>  *              or should it result in an error. This dicision may already have been made

Reviewer: ```suggestion
 *              or should it result in an error. This decision may already have been made
```

  if (container.exists()) {
    if (!isAccessAllowed(agent, container, HTTPMethod.PUT)) return accessError(agent);

    // non containment triples
    container.write(content);

    return StatusCodes.NO_CONTENT;
  }

  if (!isAccessAllowed(agent, pContainer, HTTPMethod.PUT)) return accessError(agent);

  createContainer(pContainer, container.getIri(), content);

  return StatusCodes.CREATED;
}

/**
 * agent     : The Agent trying to POST to the Container
 * container : The Container the Agent sent the request to
 * slug      : The hint provided by the requestor for the child container name
 * content   : The content for the Resource the Agent wants to POST to the Container
 *
 * Return    : 403 or 401 if the Agent does not have permission.
 *             201 if the child container is created.
 */
export function postContainer(agent: Agent, container: Container, slug: string, content: Readable) {
  if (!isAccessAllowed(agent, container, HTTPMethod.POST)) return accessError(agent);

  const iri: IRI = container.newResourceName(slug);
  createContainer(container, iri, content);

  return StatusCodes.CREATED;
}

/**

>  * agent     : The Agent tryping to PATH the container.

Reviewer: ```suggestion
 * agent     : The Agent trying to PATCH the container.
```

 * container : The Container the Agent is trying to PATCH
 * content   : The content the Agent is using to PATCH the container. This pseudo-code does

>  *             not attemptto describe how the patch operation are carried out.

Reviewer: ```suggestion
 *             not attempt to describe how the patch operation are carried out.
```

  container.applyPatch(content);

  return StatusCodes.NO_CONTENT;
}

/**
 *
 * When an Agent tries to delete a Container in the parent Container
 * with the DELETE HTTP method, the following algorithms are used to
 * enforce access control and policy propagation.
 *
 * There are 4 approaches available :
 *
 * 1. Delete the container only if it is empty
 * 2. Delete the container and its children. If any child cannot be
 *    deleted then continue to the next child. Return a list of the
 *    children that could not be deleted. This is similar to how most
 *    file systems work.
 * 3. Delete the container and its children. All children are deleted
 *    or the operation fails. This requires transactional support.
 * 4. Delete the container and its children. If the agent has permission
 *    to delete the container, then the permissions for the children
 *    are ignored and the children are deleted.
 *
 * In addition, the delete operation can be carried out synchronously or
 * asynchronously. If the client asks for the operation to be carried
 * out asynchronously and the server is capable of an asynchronous
 * operation then the response will be 202 (ACCEPTED).

>  * When the operation is complete the server will send a notification

Reviewer: ```suggestion
 * When the operation is complete, the server will send a notification
```

 * to the parent container along with the result of the operation.
 * This has not been implemented yet so there is no work done on the
 * structure of the notification or result.
 *
 * NOTE: We have not done any work on how a client would negotiate
 *       with the server on which delete mechanism to use.
 *       We could use a mechanism similar to content negotiation.
 */

/**

>  * Delete a container is it is empty.

Reviewer: ```suggestion
 * Delete a container if it is empty.
```

  const cAcr = container.getAccessControlResource();

  propagate(pAcr, pAcr.getIri(), ACP.ACCESS_MEMBERS, cAcr, ACP.ACCESS);
  propagate(pAcr, pAcr.getIri(), ACP.ACCESS_MEMBERS, cAcr, ACP.ACCESS_MEMBERS);
  propagate(pAcr, pAcr.getIri(), ACP.ACCESS_MEMBERS_PROTECTED, cAcr, ACP.ACCESS_PROTECTED);
  propagate(pAcr, pAcr.getIri(), ACP.ACCESS_MEMBERS_PROTECTED, cAcr, ACP.ACCESS_MEMBERS_PROTECTED);
  propagate(pAcr, pAcr.getIri(), ACP.ACCESS_MEMBERS_LOCKED, cAcr, ACP.ACCESS_LOCKED);
  propagate(pAcr, pAcr.getIri(), ACP.ACCESS_MEMBERS_LOCKED, cAcr, ACP.ACCESS_MEMBERS_LOCKED);

  pAcr.getAccessControls().forEach((ac: AccessControl) => {
    propagate(pAcr, ac.getIri(), ACP.APPLY_MEMBERS, cAcr, ACP.APPLY);
    propagate(pAcr, ac.getIri(), ACP.APPLY_MEMBERS, cAcr, ACP.APPLY_MEMBERS);
    propagate(pAcr, ac.getIri(), ACP.APPLY_MEMBERS_PROTECTED, cAcr, ACP.APPLY_PROTECTED);
    propagate(pAcr, ac.getIri(), ACP.APPLY_MEMBERS_PROTECTED, cAcr, ACP.APPLY_MEMBERS_PROTECTED);
    propagate(pAcr, ac.getIri(), ACP.APPLY_MEMBERS_LOCKED, cAcr, ACP.APPLY_LOCKED);
    propagate(pAcr, ac.getIri(), ACP.APPLY_MEMBERS_LOCKED, cAcr, ACP.APPLY_MEMBERS_LOCKED);
  });
}

/**

>  * Find statements matching (csubject, cpredicate, *) in the container ACR
>  * and write those statements to the resource ACR after replacing the predicate with rpredicate.

Reviewer: ```suggestion
 * Find statements matching (csubject, cpredicate, *) in the container ACR,
 * replace the predicate with rpredicate, and write those statements to the resource ACR.
```
